dbt_common/behavior_flags.py

```diff
@@ -0,0 +1,108 @@
+import inspect
+from types import SimpleNamespace
+from typing import Any, Dict, List, TypedDict
+
+try:
+    from typing import NotRequired
+except ImportError:
+    # NotRequired was introduced in Python 3.11
+    # This is the suggested way to implement a TypedDict with optional arguments
+    from typing import Optional as NotRequired
+
+from dbt_common.events.base_types import WarnLevel
+from dbt_common.events.functions import fire_event
+from dbt_common.events.types import BehaviorDeprecationEvent
+
+
+class BehaviorFlag:
+    """
+    The canonical behavior flag that gets used through dbt packages
+
+    Args:
+        name: the name of the behavior flag, e.g. enforce_quoting_on_relation_creation
+        setting: the flag setting, after accounting for user input and the default
+        deprecation_event: the event to fire if the flag evaluates to False
+    """
+
+    def __init__(self, name: str, setting: bool, deprecation_event: WarnLevel) -> None:
+        self.name = name
+        self.setting = setting
+        self.deprecation_event = deprecation_event
+
+    @property
+    def setting(self) -> bool:
+        if self._setting is False:
+            fire_event(self.deprecation_event)
+        return self._setting
+
+    @setting.setter
+    def setting(self, value: bool) -> None:
+        self._setting = value
+
+    def __bool__(self) -> bool:
+        return self.setting
+
+
+class RawBehaviorFlag(TypedDict):
+    """
+    A set of config used to create a BehaviorFlag
+
+    Args:
+        name: the name of the behavior flag
+        default:  default setting, starts as False, becomes True in the next minor release
+        deprecation_version: the version when the default will change to True
+        deprecation_message: an additional message to send when the flag evaluates to False
+        docs_url: the url to the relevant docs on docs.getdbt.com
+    """
+
+    name: str
+    default: bool
+    source: NotRequired[str]
+    deprecation_version: NotRequired[str]
+    deprecation_message: NotRequired[str]
+    docs_url: NotRequired[str]
+
+
+# this is effectively a dictionary that supports dot notation
+# it makes usage easy, e.g. adapter.behavior.my_flag
+Behavior = SimpleNamespace
```

Comment:
I would prefer to keep things strongly typed. Rather than having the RawBehaviorFlag TypedDict and this SimpleNamespace. If using the flag names as properties is important, we could implement that with __getattr__ on BehaviorFlag.

Reply:
I want to make this as easy as possible for adapter maintainers to use and I lean towards built-in types when I can (`dict` here). So I feel more strongly about the `TypedDict` part.

How would you implement the flags as properties using `getattr` on `BehaviorFlag`?

Reply:
I slept on this and I think I understand what you're saying now. The use of `SimpleNamespace` was always meant to be a placeholder until we needed something better, so that we could swap out the implementation without forcing downstream updates. Let me work something up.

Reply:
I updated Behavior to be a custom class and moved the `register` function into the `__init__` since that's effectively what it's doing. I used `__getattr__` on `Behavior` to return the flags and typed it to return only `BehaviorFlag` objects. The update to the tests both here and in `dbt-adapters` was effectively swapping `register` with `Behavior`. I'm also raising a custom exception now if the attribute (flag) is not found. What do you think of this approach?